lambeq/training/nelder_mead_optimizer.py

```diff
@@ -0,0 +1,427 @@
+# Copyright 2021-2023 Cambridge Quantum Computing Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+NelderMeadOptimizer
+=============
+Module implementing the Nelder-Mead Optimization Algorithm.
+
+"""
+from __future__ import annotations
+
+from collections.abc import Callable, Iterable, Mapping
+from typing import Any
+import warnings
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+from lambeq.core.utils import flatten
+from lambeq.training.optimizer import Optimizer
+from lambeq.training.quantum_model import QuantumModel
+
+
+class NelderMeadOptimizer(Optimizer):
+    """Nelder Mead Optimizer
+
+    The Nelder-Mead optimizer is an algorithm used for unconstrained
+    optimization in multidimensional spaces. Unlike some other
+    optimization methods, it does not take into account any bounds or
+    constraints on the variables. The algorithm is based on the Simplex
+    method and is particularly useful when the derivatives
+    (first and second)  of the objective function are unknown or
+    unreliable.
+
+    Although the Nelder-Mead algorithm is generally robust and widely
+    applicable, it has some limitations. In cases where the derivatives
+    can be accurately computed, alternative algorithms that utilize
+    this derivative information may offer better performance. These
+    methods are often preferred due to their ability to handle a wider
+    range of scenarios and their tendency to converge to more optimal
+    solutions. It is worth noting that the Nelder-Mead technique is a
+    heuristic search approach, which means that it can sometimes
+    converge to non-stationary points or suboptimal solutions.
+
+    This implementation is based heavily scipy's optimize.minimize. See
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html
+
+    """
+
+    model: QuantumModel
+
+    def __init__(
+        self,
+        model: QuantumModel,
+        hyperparams: dict[str, float],
+        loss_fn: Callable[[Any, Any], float],
+        bounds: ArrayLike | None = None,
+    ) -> None:
+        """Initialise the Nelder-Mead optimizer.
+
+        The hyperparameters may contain the following key value pairs:
+
+        - `adaptive`: Adjust the algorithm's parameters based on the
```

Comment:
This is a lot better, but I think it would be nice if you put the type and default at the start here too, perhaps similar to documenting the parameters, like:
```
- adaptive : int, default: 1000
    Adjust the ...
```

Indenting by only 4 should make the remaining width more bearable to work with too